### PR TITLE
fix: make properties of configurables optional in typegen

### PR DESCRIPTION
**Pull request by Torres-ssf**
- Closes https://github.com/FuelLabs/fuels-ts/issues/2790

# Summary

This PR updates the typegen template to ensure that all configurable properties are optional.

# Checklist

- [x] I **_added_** — `tests` to prove my changes
- [ ] I **_updated_** — all the necessary `docs`
- [x] I **_reviewed_** — the entire PR myself, using the GitHub UI
- [ ] I **_described_** — all breaking changes and the Migration Guide

### CODE REVIEW

**Review by arboleya — DISMISSED** [review]

File: packages/fuel-gauge/src/policies.test.ts [hunk 1]
```diff
@@ -157,8 +157,8 @@ describe('Policies', () => {
     } = launched;
 
     const factory = new ContractFactory(
-      ScriptMainArgsAbi__factory.bin,
-      ScriptMainArgsAbi__factory.abi,
+      PayableAnnotationAbiHex,
```
Comment (maschad) [inline packages/fuel-gauge/src/policies.test.ts:160]: Why were these changed to `PayableAnnotation`  abi?

File: packages/fuel-gauge/src/predicate/predicate-configurables.test.ts [hunk 2]
```diff
@@ -198,6 +198,50 @@ describe('Predicate', () => {
...
+        wallets: [wallet],
+      } = launched;
+
+      const configurableConstants = {
+        ADDRESS: getRandomB256(),
+      };
+
+      const amountToTransfer = 300;
+
+      const predicate = new Predicate({
```
Comment (maschad) [inline packages/fuel-gauge/src/predicate/predicate-configurables.test.ts:215]: I think given we are testing out the typegen for this use case we should probably use the `createInstance` syntax to demonstrate the type system is accommodating the partiality.  e.g.

```ts
const predicate = PredicateWithConfigurableAbi__factory.createInstance(
  provider,
  [myFee, myAddress ],
  { FEE: 234 }
);
```

**Review by maschad — CHANGES_REQUESTED** [review]
Looks good, left two comments

File: packages/fuel-gauge/src/policies.test.ts (hunk 1 above)
Comment (petertonysmith94) [reply packages/fuel-gauge/src/policies.test.ts:160]: I assume because the `ScriptMainArgs` is not a contract and `PayableAnnotation` is.

Comment (maschad) [reply packages/fuel-gauge/src/policies.test.ts:160]: That's valid 😅 , perhaps it should be a script instance, although `ContractFactory` was used from this [test's inception](https://github.com/FuelLabs/fuels-ts/pull/1437/files#diff-0c6d881187bb1fdc9547309f7ee2830612689983980df36e3ec61be1224670b2R95), perhaps it was an oversight then as well. @Torres-ssf thoughts?

Comment (Torres-ssf) [reply packages/fuel-gauge/src/policies.test.ts:160]: @maschad I am using `PayableAnnotation` because the `ScriptMainArgs` ABI and bytes were being incorrectly used as a contract ABI and bytes.

Comment (maschad) [reply packages/fuel-gauge/src/policies.test.ts:160]: I suppose it doesn't really matter whether this is contract or a script for the purposes of this test, which is why it was passing before.

File: packages/fuel-gauge/src/predicate/predicate-configurables.test.ts (hunk 2 above)
Comment (Torres-ssf) [reply packages/fuel-gauge/src/predicate/predicate-configurables.test.ts]: https://github.com/FuelLabs/fuels-ts/pull/2815/commits/0242f3bd6ae7a9bd9bf455ddfc07c6052113cc62

**Review by maschad — DISMISSED** [review]
Nice

File: packages/fuel-gauge/src/policies.test.ts (hunk 1 above)
Comment (Torres-ssf) [reply packages/fuel-gauge/src/policies.test.ts:160]: @maschad The tests were passing because in both examples, the contract was never being deployed.

**Review by petertonysmith94 — DISMISSED** [review]

File: packages/fuel-gauge/src/policies.test.ts (hunk 1 above)
Comment (nedsalk) [reply packages/fuel-gauge/src/policies.test.ts:160]: This will become impossible once #2787 / #989 is done.

**Review by nedsalk — APPROVED** [review]

File: .changeset/tasty-fireants-switch.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@fuel-ts/abi-typegen": patch
+---
+
+fix: make Predicate configurables option in typegen
```
Comment (nedsalk) [inline .changeset/tasty-fireants-switch.md:5]:
```suggestion
fix: make configurables properties optional in typegen
```

**Review by nedsalk — CHANGES_REQUESTED** [review]
This PR currently only solves the problem for predicates, but #2790 is a general issue that includes scripts and contracts as well, so they should be covered in this PR also.

File: .changeset/tasty-fireants-switch.md (hunk 3 above)
Comment (Torres-ssf) [reply .changeset/tasty-fireants-switch.md:5]: @nedsalk Thanks for your suggestion. 

I am having a hard time understanding how this issue extends to Contracts and Scripts. 

Could you please provide a detailed outline of the error and an example of a scenario where this issue occurs for Contracts and Scripts?

Comment (nedsalk) [reply .changeset/tasty-fireants-switch.md:5]: For contracts, you would have to change [this line](https://github.com/FuelLabs/fuels-ts/blob/8c5171d947291814c6bca0334d72379e51167167/packages/abi-typegen/src/templates/contract/dts.hbs#L44) to include the `Partial<...>` as you did for the predicates. The relevant script file to change is [this one](https://github.com/FuelLabs/fuels-ts/blob/8c5171d947291814c6bca0334d72379e51167167/packages/abi-typegen/src/templates/script/factory.hbs#L47). Upon changing it, you'll find that some tests are failing and you'll have to do a change like you did [here](https://github.com/FuelLabs/fuels-ts/pull/2815/files#diff-50b9975b1c89e23c847d731fb5c02edaac999b537242e2056e29451e904cae3fR19).

Comment (Torres-ssf) [reply .changeset/tasty-fireants-switch.md:5]: @nedsalk Ok so for Scripts and Contracts this only results in error if you use the generated configurable type.

That is why I did not understand the request.

The error is flagged for predicates as soon as you try to instantiate the predicate with a partial configurable.

But for Scripts and Contracts it only happens if you type safe the configurable object:

```ts
const configurable: ConfigurableContractAbiConfigurables {
   ...
}
```

Comment (nedsalk) [reply .changeset/tasty-fireants-switch.md:5]: Okay, so configurables for contracts and scripts aren't taken into account in the generated factory code. This is one more reason to go for #2686 which will make it easy to add this type safety.

**Review by nedsalk — APPROVED** [review]
